src/core/connection.cpp

```diff
@@ -33,360 +33,631 @@
-        if (messages.empty())
-            break;
+                case Topic::AUTH:
+                    handle_authentication_response(parsed_message);
+                    break;
 
-        for (const Message& message : messages) {
-            if (message.topic() == Topic::CONNECTION && message.action() == Action::PING) {
-                MessageBuilder pong(Topic::CONNECTION, Action::PONG);
-                send_(pong);
-                continue;
+                default:
+                    {
+                        std::stringstream error_message;
+                        error_message << "unsolicited message: " << parsed_message.header();
+                        error_handler_.on_error(error_message.str());
+                        assert(0);
+                    }
             }
 
-            switch (message.topic()) {
-            case Topic::EVENT:
-                p_event->notify_(message);
-                break;
+        }
+    }
+
+    void Connection::handle_connection_response(const Message &message)
+    {
+        ConnectionState new_state = transition_incoming(state_, message);
 
-            case Topic::PRESENCE:
-                p_presence->notify_(message);
-                break;
+        if (new_state == ConnectionState::ERROR) {
+            std::stringstream error_message;
+            error_message << "connection error state reached, previous state: " << state_
+                << ", message header: " << message.header();
+            error_handler_.on_error(error_message.str());
+        }
 
+        state(new_state);
+        switch(message.action()) {
+            case Action::PING:
+                {
+                    const MessageBuilder pong(Topic::CONNECTION, Action::PONG);
+                    send(pong);
+                } break;
+            case Action::CHALLENGE:
+                {
+                    MessageBuilder challenge_response(Topic::CONNECTION, Action::CHALLENGE_RESPONSE);
+                    challenge_response.add_argument(ws_handler_.URI());
+                    send(challenge_response);
+                } break;
+            case Action::CHALLENGE_RESPONSE:
+                {
+                    if (p_auth_params_) {
+                        send_authentication_request();
+                    }
+                } break;
+            case Action::REDIRECT:
+                {
+                    if (message.num_arguments() < 1) {
+                        error_handler_.on_error("No URI given in connection redirect message");
+                        break;
+                    }
+                    const Buffer &uri_buff(message[0]);
+                    ws_handler_.URI(std::string(uri_buff.cbegin(), uri_buff.cend()));
+                } break;
             default:
-                assert(0);
-                close();
-                return;
-            }
+                {
+                    std::stringstream error_message;
```

Comment:
doesn't the message parser say theres a parse error if the connection message (including action) is invalid?